client/src/main/java/io/pravega/client/segment/impl/SegmentInputStreamImpl.java

```diff
@@ -164,12 +169,25 @@ private void handleRequest() {
         }
     }
 
+    private void verifyIsAtCorrectOffset(WireCommands.SegmentRead segmentRead) {
+        long offsetRead = segmentRead.getOffset() + segmentRead.getData().position();
+        long expectedOffset = offset + buffer.dataAvailable();
+        checkState(offsetRead == expectedOffset, "ReadSegment returned data for the wrong offset %s vs %s", offsetRead,
+                   expectedOffset);
+    }
+
     /**
      * @return If there is enough room for another request, and we aren't already waiting on one
      */
     private void issueRequestIfNeeded() {
-        if (!receivedEndOfSegment && outstandingRequest == null && buffer.capacityAvailable() > readLength) {
-            outstandingRequest = asyncInput.read(offset + buffer.dataAvailable(), readLength);
+        if (!receivedEndOfSegment && buffer.capacityAvailable() > readLength) {
+            if (outstandingRequest == null) {
+                outstandingRequest = asyncInput.read(offset + buffer.dataAvailable(), readLength);
+            } else if (outstandingRequest.isCompletedExceptionally()) {
+                Throwable e = FutureHelpers.getException(outstandingRequest);
+                log.warn("Encountered an exception while reading for " + asyncInput.getSegmentId(), e);
```

Comment:
I have to do that due to the way the arguments work to the log method. There is no way to do formatting and log an exception in the same log line